src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::credentials::Result;
+use crate::credentials::internal::sts_exchange::ClientAuthentication;
+use crate::errors;
+use crate::headers_util::build_bearer_headers;
+use crate::token::{CachedTokenProvider, Token, TokenProvider};
+use crate::token_cache::TokenCache;
+use gax::error::CredentialsError;
+use http::{Extensions, HeaderMap};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::time::{Duration, Instant};
+
+use super::Credentials;
+use super::dynamic::CredentialsProvider;
+use super::external_account_sources::url_sourced_account::UrlSourcedCredentials;
+use super::internal::sts_exchange::{ExchangeTokenRequest, STSHandler};
+
+#[async_trait::async_trait]
+pub(crate) trait SubjectTokenProvider: std::fmt::Debug + Send + Sync {
+    /// Generate subject token that will be used on STS exchange.
+    async fn subject_token(&self) -> Result<String>;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceFormat {
+    #[serde(rename = "type")]
+    pub format_type: String,
+    pub subject_token_field_name: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSourceHeaders {
+    #[serde(flatten)]
+    pub headers: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CredentialSource {
+    pub url: Option<String>,
+    pub headers: Option<CredentialSourceHeaders>,
+    pub format: Option<CredentialSourceFormat>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ExternalAccountConfig {
+    pub audience: String,
+    pub subject_token_type: String,
+    pub token_url: String,
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub credential_source: CredentialSource,
+}
+
+#[derive(Debug)]
+struct ExternalAccountTokenProvider {
+    subject_token_provider: Box<dyn SubjectTokenProvider>,
+    config: ExternalAccountConfig,
+}
```

Comment:
Depending one why you are using `dyn` here, it may be more idiomatic to use:

```rust
#[derive(Debug)]
struct ExternaAccountTokenProvider<T> {
  subject_token_provider: T,
  config: ExternalAccountConfig,
}
```

and then;

```rust
impl<T> TokenProvider for ExternalAccountProvider<T> where T: SubjectTokenProvider {
  // as below
```

The point being that in Rust generics are the way to go.  The difficulty might with any function that creates different subject providers based on an external configuration file.  In a vacuum, I would say one level of `dyn` should be enough.